### PR TITLE
bench(#632): shard-only reach benchmark — dense CTM-AD value_and_grad D=10-12

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @Ying-Jer Kao.

Docs + example only (no src/test change). Follow-up to the #632 Increment-2 chunked-backward NO-GO (in #668): since the large-D backward memory lever is GSPMD sharding (rung-2), this measures how far *that* lever reaches at D=10-12 with the full implicit-AD `value_and_grad`.

**Harness validated:** 1-GPU D=8 χ=24 = 10.66 GB — matches the rung-2 gate 2B number exactly (uses the rung-2 probe, recipe=2×2).

**Result — on 2 A100s, sharding gives NO reach extension at D=10-12.** Relief is χ-gated and D-fading:

| D | χ=24 (1→2 GPU) | χ=16 (1→2 GPU) |
|---|---|---|
| 6 | 2.16→1.15 GB (1.88×) | 0.81→0.91 GB (0.89×) |
| 8 | 10.66→6.85 GB (1.56×) | 5.10→4.84 GB (1.05×) |
| 10 | OOM / OOM | 23.21→23.20 GB (1.00×) |
| 12 | autotune-fail / OOM | OOM / OOM (2-GPU +77 vs 1-GPU +30.7 GiB) |

Every D≥10 config either fits on 1-GPU (D10 χ16) or OOMs on both (D10 χ24, all D12); D=12 sharding is counterproductive (all-gather buffers for the replicated intermediate). This confirms the SVD-replication root cause at large D: the projector SVD forces the χ²·D⁶ backward intermediate replicated, so ÷N shards only the contraction. Neither chunk (#668 Inc-2 NO-GO) nor 2-GPU shard beats the SVD-bound backward.

**Caveat:** only 2 clean A100s at run time; rung-2's +2-in-D used 4 GPUs. The relief-fade + D=12 overhead predict D=12 out of reach even at 4-GPU, but that is untested. Full findings: `docs/superpowers/handoffs/2026-07-01-632-shard-only-reach-grad-D10-12.md`.

### CODE REVIEW

File: examples/bench_ctm_shard_reach_grad.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+def peak_gb():
+    try:
+        return jax.devices()[0].memory_stats()["peak_bytes_in_use"] / 1e9
+    except Exception:
+        return float("nan")
+
+
+def main():
+    ap = argparse.ArgumentParser()
+    ap.add_argument("--D", type=int, nargs="+", default=[10, 12])
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply examples/bench_ctm_shard_reach_grad.py:56]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @Ying-Jer Kao.

Fixed in 2010a4d. A multi-`--D` invocation (e.g. the default `--D 10 12`) now re-execs one fresh subprocess per D, so each reported `per_device_peak` is an uncontaminated process high-water mark; single-`--D` runs are unchanged. Confirmed the contamination is real today on unrelated HOTRG work — a combined single+shard process masked the sharded peak with the earlier config's high-water. Docstring updated to document the per-process behavior.

File: docs/superpowers/handoffs/2026-07-01-632-shard-only-reach-grad-D10-12.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+1× (and below). This is the *same* wall the chunked-backward NO-GO hit from the
+other side: neither chunk nor shard addresses the SVD-replicated backward
+intermediate — that is the structural D≥10 dense CTM-AD wall.
+
+## Limitation / open
+
+- **Only 2 clean A100s at run time** (GPU 0 held 61 GiB; GPU 4 at 96% util). rung-2
+  reported 4-GPU extends the single ceiling D=8 → D=10 at χ=24 (+2 in D, 21.45
+  GB/device). 2-GPU is insufficient to reach D=10 χ=24 (both OOM). A 4-GPU rerun
+  (`CUDA_VISIBLE_DEVICES=0,1,2,4` or any 4 free A100s) would test whether the +2-in-D
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply docs/superpowers/handoffs/2026-07-01-632-shard-only-reach-grad-D10-12.md]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @Ying-Jer Kao.

Fixed in 2010a4d. You're right that a bare `CUDA_VISIBLE_DEVICES=0,1,2,4` is ordering-dependent. Verified on the box today: with `CUDA_DEVICE_ORDER=PCI_BUS_ID` the DGX display GPU is PCI index **3** (4 GB) and indices **0,1,2,4** are the four A100s, so the correct 4-GPU mask is `CUDA_DEVICE_ORDER=PCI_BUS_ID CUDA_VISIBLE_DEVICES=0,1,2,4 NCCL_P2P_DISABLE=1`. A plain 4-way all-reduce over those four devices ran cleanly (no NCCL stall) and a 4-GPU HOTRG run gave bit-parity with single-device. The handoff doc's "Limitation / open" now states this explicitly (and warns a shared/busy A100 OOMs its shard and stalls the clique). The example script gained the same verified invocation.
